### ISSUE TITLE
Disabled buttons on reach end or start

### ISSUE THREAD

**Issue by dngraphisme** [barthy-koeln/scroll-snap-slider#15]
Hello Barthy,

Thanks for your library :)

I was looking for this kind of library for a while to not use swiperJS for overflow-x

Here is my code, i managed to create multiple instances for multiple "carousel" :

```
$(".slider-snap > div").each(function () {
            let $this = $(this);
            const sliderElement = this;
            const slides = sliderElement.getElementsByClassName('scroll-snap-slide');
            const slider = new ScrollSnapSlider(sliderElement, true, [
                //new ScrollSnapAutoplay(5000),
                new ScrollSnapDraggable(50)
            ]);

            // navigation
            const prev = $this.find('.navigation .prev');
            const next = $this.find('.navigation .next');

            // event on scroll overflow-x
            const setSelected = function (event) {
                const slideElementIndex = event.detail
                //const slideElement = slides[slideElementIndex]
                //console.log(slideElementIndex)

                if (slideElementIndex == 0) {
                    prev.addClass("disabled");
                } else {
                    prev.removeClass("disabled");
                }

                if (slideElementIndex == numberPages() - 1) {
                    next.addClass("disabled");
                } else {
                    next.removeClass("disabled");
                }

            }
            slider.addEventListener('slide-pass', setSelected)
            slider.addEventListener('slide-stop', setSelected)

            // calc number pages
            function numberPages() {
                let $scrollSnapSlider = Number($('.scroll-snap-slide:nth-child(1)').css('flex-basis').slice(0, -1));
                let $scrollSlides = $('.scroll-snap-slide').length;
                let $slidesPerPage = Math.round(100 / $scrollSnapSlider); // nbr de slides
                let $numberPage = Math.ceil($scrollSlides / $slidesPerPage);
                //console.log($numberPage)
                return ($numberPage);
            }

            let page = 1;

            prev.on('click', function () {
                slider.slideTo(slider.slide - 1)
                page -= 1;
                if (page <= 1) {
                    page = 1
                    prev.addClass("disabled");
                } else {
                    prev.removeClass("disabled");
                }
                next.removeClass("disabled");
                //console.log('prev', page)
            });

            next.on('click', function () {
                slider.slideTo(slider.slide + 1)
                page += 1;
                if (page >= numberPages()) {
                    page = numberPages();
                    next.addClass("disabled");
                } else {
                    next.removeClass("disabled");
                }
                prev.removeClass("disabled");
                //console.log('next', page)
            });

        });
```

There is a problem with the next button, I can't get the correct calculation.

console.log(slideElementIndex) > this gives me the wrong number compared to my numberPages() function

There is a problem I think with the scroll-snap-slide

Do you have an idea ?

Thank you for your help :)

**Comment by barthy-koeln** [comment]
Hi Nicko,

**Comment by dngraphisme** [comment]
Hello @barthy-koeln thanks for your feedback.

Here is a link for a codepen sandbox: https://codepen.io/dngraphisme/pen/ZErGdXd

Next: the numberPages() function allows me to calculate the number of pages. What I call page is going to be the number of "scroll-snap-slide" per view.

I am looking to put a disabled class following the positioning of "scroll-snap-slide".

I get there with the prev button but the next button depending on the number of "scroll-snap-slide" I completely bugged. Worse when it's an odd number... Maybe I bothered to do this function for nothing ^^'

I don't understand when you tell me that I can have the end and the beginning with `slides.length - 1`

Thanks, Nickolas

**Comment by barthy-koeln** [comment]
Thanks for the example, that cleared things up for me. This library currently only supports slides, that have 100% width of the slider.

This is stated in the README file:

> Slides always have 100% width. You can add whatever markup inside.

This means that you can include multiple images in one slide, but they will always act as a single full-width slide.

Here's an example with multiple images per slide and some JS to show/hide elements based on the position:

https://codepen.io/BarthyB/pen/rNJVXGy

**Comment by dngraphisme** [comment]
Hello @barthy-koeln , ok i understand.

Do you plan to improve this functionality for once?

Thanks again for your quick feedback.

**Comment by barthy-koeln** [comment]
Good morning!

I doubt that I will spend much time, as I have lots of To-Dos already. But I will try around today and see if I find an easy solution.
Can you describe the feature you would like to see in two or three sentences?

"I need a slider, that ..."

**Comment by barthy-koeln** [comment]
Hello again,

Please check out the branch `feature/slide-smaller-than-slider`. You can clone or download the project and open the index.html with any local server.

The demo in that branch should have a slider that shows multiple slides at once, yet only slides one at a time.

The main differences in the code are:

```css
.example-slider {
  /* [...] */
  width: 100%;
  padding: 0 calc(50vw - 200px);
  box-sizing: border-box;
}

.example-slider .scroll-snap-slide {
  width: 400px;
  scroll-snap-align: center;
}
```

Note that I add half the size of a slide plus half the size of the window (the slider's container) as horizontal padding to the slider.

```js
// ...
this.sizingMethod = () => {
  return this.element.firstElementChild.offsetWidth
}
// ...
```

I calculate the width of a slide like this. You can override this method if necessary. All slides should have the same width, though.

___

I also added some code to enable/disable the arrows.
The plugins do not all work (only autoplay should work as expected).

Please test around and report any issues! If we can find a good solution, I will incorporate it in the library.

**Comment by dngraphisme** [comment]
Wow @barthy-koeln you are awesome ^^... I test it asap and come back to you. 

Thank you very much !

**Comment by barthy-koeln** [comment]
Pre-Edit: I actually forgot to push the latest changes. Please pull or re-download the same branch and try the demo again!
___

The padding at the left/right are important for the calculations, since it allows the first and last slides to be centred in the slider. Does the demo without any modifications work as expected?

Without dot pagination and without extra padding, I have an idea how to check if the scrolling element has reached the end. I will try ASAP, at the latest on Monday.

**Comment by dngraphisme** [comment]
Hello @barthy-koeln ,

I'll get back to you :

**Yes for your demo everything works !**

Now for the padding I don't agree with this principle because if you use a framework, the but is to keep the grid intact for the responsive... otherwise it will be unmanageable with @media to manage the whole thing.

On my site that I am currently developing, I explain to you how I managed this:

I put ".scroll-snap-slider" in a col-xl-8 and added offset-xl-4 to snap the element to the right of the viewport. (https://ibb.co/2FjK02D)

Now the problem I have is to manage the overflow of objects.

If my ".scroll-snap-slide" are in flex: 0 0 48%, I always have this problem where a thumbnail will be truncated (border on the left not visible) This only happens on drag and not on swipe or button > https://ibb.co/3hVMPm2

And then otherwise I still have the problem of the arrow next which does not go into disabled but as you mentioned you will try to check this problem.

Thanks again

**Comment by barthy-koeln** [comment]
How you manage the padding/offset is totally up to you. My examples use pixel values for simplicity, but the bootstrap offset classes are a good fit.
___
I just pushed the `feature/scroll-end-not-in-center` branch. This uses `scroll-snap-align: start;`, no padding, and the correct calculations for the arrow buttons.

Is that the behavior you are looking for?
___

I can't replicate your issue with the missing border. If it persists, I might need another code example.

**Comment by dngraphisme** [comment]
@barthy-koeln ok thanks i check this monday ;)

**Comment by barthy-koeln** [comment]
Hi @dngraphisme 

I did not receive any notification about your edit, next time feel free to create a new message in this thread! I will receive an email and a notification then.

I will send out a new NPM release very soon (this week). I need to document the changes properly and update the demo first.

The responsiveness feature that you describe sounds to me like something specific for each project. Therefore, I would like to keep it out of the core library. I will also add an example on how to use the constructor's second parameter,  `attachEventListeners` and `destroy` in this case, though. That's what they are there for!

**Comment by dngraphisme** [comment]
Hello @barthy-koeln oh sorry for the notification ^^'

Ok great for NPM thanks !

I'm also waiting for your example for these parameters: attachEventListeners and destroy

I could modify my code accordingly :)

Let me know when you've done this.

thanks again

**Comment by dngraphisme** [comment]
Hello @barthy-koeln , i have a last question :

How can i detect and launch arrows depending on the view of ".scroll-snap-slide"

Let's take an example:

On desktop I have 12 elements so not scrollable with the library
On iPad, the 12 elements become 8 visible with 4 hidden.

Do you have an idea to make the arrows visible only when you need to swipe to the right ?

Thanks again, you're the best :)

**Comment by barthy-koeln** [comment]
You would have to add a resize listener on the `window`, that checks if the `sliderElement.scrollWidth` is bigger than `sliderElement.clientWidth`. If it is, your slider can slide. I will include this in the demo/plugin.

**Comment by dngraphisme** [comment]
@barthy-koeln ok great ! Thanks :)

I also noticed one thing, the '.croll-snap-slide' don't always start completely on the left.

On your side too ?

**Comment by barthy-koeln** [comment]
Hi Nicko,

I just released v2.5.0.

Take a look at the [updated demo page](https://barthy-koeln.github.io/scroll-snap-slider/)! It includes a slider with multiple slides and on/off arrows, as well as a responsiveness example.

You can find the code in the [`demo` folder](https://github.com/barthy-koeln/scroll-snap-slider/tree/main/demo).
___

I will try around with flex properties, especially with bootstrap's negative `.row` margins...
If you want to submit a full bug report with a replicable example, please open a new Issue :)
